pageserver/src/tenant/storage_layer/image_layer.rs

```diff
@@ -943,11 +959,112 @@ impl Drop for ImageLayerWriter {
     }
 }
 
+pub struct ImageLayerIterator<'a, 'ctx> {
+    image_layer: &'a ImageLayerInner,
+    ctx: &'ctx RequestContext,
+    next_batch_start_key: Key,
+    key_values_batch: Vec<(Key, Lsn, Value)>,
+    next_idx_in_batch: usize,
+    is_end: bool,
+    batch_size: usize,
+}
+
+impl<'a, 'ctx> ImageLayerIterator<'a, 'ctx> {
+    /// Retrieve a batch of key-value pairs into the iterator buffer with a single (vectored) I/O.
+    async fn next_batch(&mut self) -> anyhow::Result<()> {
+        assert!(self.next_idx_in_batch >= self.key_values_batch.len());
+        assert!(!self.is_end);
+        self.key_values_batch.clear();
+        self.next_idx_in_batch = 0;
+
+        let block_reader = FileBlockReader::new(&self.image_layer.file, self.image_layer.file_id);
+        let tree_reader = DiskBtreeReader::new(
+            self.image_layer.index_start_blk,
+            self.image_layer.index_root_blk,
+            &block_reader,
+        );
+        let mut search_key = [0; KEY_SIZE];
+        self.next_batch_start_key
+            .write_to_byte_slice(&mut search_key);
```

Comment:
the docs of this function say you should not use it for metadata keys. Does this mean that the iterator can't be used for the image layer that uses metadata keys? And if we use this encoding for metadata keys then maybe the docs of that function should be adjusted.

Reply:
oh, it means that if the user wants to get the 16B metadata key, they should not use this function, because it always produce a 18B key with some extra data.

Reply:
but here we expect a 18B key, which is universal across data/metadata keyspaces.

Reply:
refined the doc to mention it's not used to get 16B metadata keys (but it's fine to use it in general to get a 18B key for all keyspaces) https://github.com/neondatabase/neon/pull/8006/commits/0dba1f8b6ddbd3a7d11bf7612af51f33713d12c3